shared/src/main/scala/async/channels.scala

```diff
@@ -272,39 +323,43 @@ object ChannelMultiplexer:
   private enum Message:
     case Quit, Refresh
 
-  def apply[T]()(using ac: Async): ChannelMultiplexer[T] = new ChannelMultiplexer[T]:
+  def apply[T](): ChannelMultiplexer[T] = new ChannelMultiplexer[T]:
     private var isClosed = false
     private val publishers = ArrayBuffer[ReadableChannel[T]]()
     private val subscribers = ArrayBuffer[SendableChannel[Try[T]]]()
-    private val infoChannel: BufferedChannel[Message] = BufferedChannel[Message](1)
-    ac.scheduler.execute { () =>
+    private val infoChannel = UnboundedChannel[Message]()
+
+    def start()(using Async) = {
       var shouldTerminate = false
-      var publishersCopy: List[ReadableChannel[T]] = null
+      var publishersCopy: Seq[ReadableChannel[T]] = null
       var subscribersCopy: List[SendableChannel[Try[T]]] = null
       while (!shouldTerminate) {
         ChannelMultiplexer.this.synchronized:
-          publishersCopy = publishers.toList
-
-        val got = Async.await(Async.either(infoChannel.canRead, Async.race(publishersCopy.map(c => c.canRead)*)))
-        got match {
-          case Left(Success(Message.Quit)) => {
-            ChannelMultiplexer.this.synchronized:
-              subscribersCopy = subscribers.toList
-            for (s <- subscribersCopy) s.send(Failure(channelClosedException))
-            shouldTerminate = true
-          }
-          case Left(Success(Message.Refresh)) => ()
-          case Left(Failure(_))               => shouldTerminate = true // ?
-          case Right(v) => {
-            ChannelMultiplexer.this.synchronized:
-              subscribersCopy = subscribers.toList
-            var c = 0
-            for (s <- subscribersCopy) {
-              c += 1
-              s.send(v)
-            }
-          }
-        }
+          publishersCopy = publishers.toSeq
+
+        Async.select(
+          (infoChannel.readSource ~~> {
+            case Left(_) | Right(Message.Quit) =>
+              ChannelMultiplexer.this.synchronized:
+                subscribersCopy = subscribers.toList
+              for (s <- subscribersCopy) s.send(Failure(ChannelClosedException()))
+              shouldTerminate = true
+            case Right(Message.Refresh) => ()
```

Comment:
Why does this exist? And if it is only quit left, then I guess the info channel is not needed anymore. You can implement the quit using a (classical) Promise instead.

Reply:
`Refresh` is needed as like a `continue` for the while loop. Happens when we made a change to the subscriber/publisher list.